Bulk Form Management (archive/restore)

sravfeyn opened the pull request:
Implements https://docs.google.com/document/d/1XjdGI3dv11JpfIDSiU1hn_g75rnRDTY8Y6FyLBTV4fE/edit

sravfeyn committed:
experiment

sravfeyn committed:
Merge master

sravfeyn committed:
add archived_forms ES filter

sravfeyn committed:
split submitHistory report into mixin and report

sravfeyn committed:
pull XForm filter out

sravfeyn committed:
Make Bulk-archive-forms use SubmitHistoryMixin

sravfeyn committed:
actually make the template work

sravfeyn committed:
Make bulk-archive/restore an Async task

sravfeyn committed:
Handle select-all XForms - just skeleton

sravfeyn committed:
don't archive/restore already archived/restored forms

sravfeyn committed:
get bulk-archive/restore working

sravfeyn committed:
merge master into FormArchive

sravfeyn committed:
pass es-query if all filtered rows are selected

sravfeyn committed:
try using same archive function with couch,es iterators

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
Revert "try using ES iterator to archive"

This reverts commit c0fda2941c0b591071f3ce1796c998c85975aefd.

sravfeyn committed:
fix incorrect ret

sravfeyn committed:
make All-forms ES work, refactor archive/restore mode into 1 place

sravfeyn committed:
correct restore/archive feedback

sravfeyn committed:
show progress graphic correctly

sravfeyn committed:
better UX

sravfeyn committed:
refactor common logic into single class

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
remove todo, prints, debugs

sravfeyn committed:
make archive_forms task work with excel as it was

sravfeyn committed:
remove now unused exception

sravfeyn committed:
move code around, comments

sravfeyn committed:
translation

sravfeyn committed:
use form-filter-url instead of passing es-query around

sravfeyn committed:
Move bulk-form-management to standard plan

sravfeyn committed:
Merge branch 'master' into FormArchive

dimagimon commented:
There are 5 errors:

* corehq/apps/data_interfaces/interfaces.py, line 18 - F401 'BaseReportFilter' imported but unused
* corehq/apps/data_interfaces/views.py, line 33 - F401 'HttpResponseBadRequest' imported but unused
* corehq/apps/data_interfaces/views.py, line 478 - E226 missing whitespace around arithmetic operator
* corehq/apps/reports/display.py, line 41 - E226 missing whitespace around arithmetic operator
* corehq/apps/reports/standard/inspect.py, line 38 - E241 multiple spaces after ','

sravfeyn committed:
fix excel archive tests

dimagimon commented:
There are 5 errors:

* corehq/apps/data_interfaces/interfaces.py, line 18 - F401 'BaseReportFilter' imported but unused
* corehq/apps/data_interfaces/views.py, line 33 - F401 'HttpResponseBadRequest' imported but unused
* corehq/apps/data_interfaces/views.py, line 478 - E226 missing whitespace around arithmetic operator
* corehq/apps/reports/display.py, line 41 - E226 missing whitespace around arithmetic operator
* corehq/apps/reports/standard/inspect.py, line 38 - E241 multiple spaces after ','

sravfeyn committed:
pep8

sravfeyn committed:
Merge branch 'master' into FormArchive

dimagimon commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -450,3 +456,83 @@ def post(self, request, *args, **kwargs):
+    def post(self, request, *args, **kwargs):
+        form_ids_or_url = self._get_form_ids_or_url(request)
+        mode = self.request.POST.get('mode')
+        task_ref = expose_cached_download(None, expiry=1*60*60)
```
E226 missing whitespace around arithmetic operator

dimagimon commented on corehq/apps/reports/display.py:
```diff
@@ -10,6 +20,50 @@ def replace(self, *args):
+        username = self.form["form"]["meta"].get("username")
+        try:
+            if username not in ['demo_user', 'admin']:
+                full_name = get_cached_property(CouchUser, uid, 'full_name', expiry=7*24*60*60)
```
E226 missing whitespace around arithmetic operator

sravfeyn committed:
don't check for permissions when passing from task

sravfeyn committed:
pass just the query string, not the whole URL

dimagimon commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -450,3 +456,83 @@ def post(self, request, *args, **kwargs):
+    def post(self, request, *args, **kwargs):
+        form_ids_or_query_string = self.get_form_ids_or_query_string(request)
+        mode = self.request.POST.get('mode')
+        task_ref = expose_cached_download(None, expiry=1*60*60)
```
E226 missing whitespace around arithmetic operator

sravfeyn committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into FormArchive

sravfeyn committed:
datamigration for new bulk_form_management privilege

Commit:
Merge branch 'master' into FormArchive

Commit:
move migration to south_migrations

dimagimon commented:
There are 78 errors:

* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 17 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 19 - E501 line too long (131 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 20 - E501 line too long (145 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 21 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 22 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 23 - E501 line too long (127 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 30 - E501 line too long (153 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 34 - E501 line too long (163 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 36 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 40 - E501 line too long (123 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 42 - E501 line too long (122 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 43 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 45 - E501 line too long (124 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 50 - E501 line too long (137 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 53 - E501 line too long (143 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 60 - E501 line too long (135 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 61 - E501 line too long (150 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 62 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 64 - E501 line too long (159 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 66 - E501 line too long (162 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 68 - E501 line too long (172 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 70 - E501 line too long (215 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 75 - E501 line too long (150 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 76 - E501 line too long (136 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 77 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 78 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 82 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 83 - E501 line too long (185 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 91 - E501 line too long (141 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 100 - E501 line too long (145 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 112 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 113 - E501 line too long (143 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 117 - E501 line too long (138 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 119 - E501 line too long (140 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 123 - E501 line too long (136 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 124 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 128 - E501 line too long (126 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 134 - E501 line too long (153 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 135 - E501 line too long (136 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 139 - E501 line too long (138 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 141 - E501 line too long (168 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 143 - E501 line too long (143 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 145 - E501 line too long (176 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 147 - E501 line too long (138 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 152 - E501 line too long (125 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 153 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 156 - E501 line too long (134 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 157 - E501 line too long (123 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 161 - E501 line too long (135 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 162 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 165 - E501 line too long (156 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 179 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 180 - E501 line too long (168 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 184 - E501 line too long (145 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 185 - E501 line too long (176 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 197 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 201 - E501 line too long (138 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 202 - E501 line too long (150 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 206 - E501 line too long (123 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 209 - E501 line too long (128 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 213 - E501 line too long (150 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 215 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 224 - E501 line too long (160 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 225 - E501 line too long (119 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 226 - E501 line too long (135 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 227 - E501 line too long (116 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 228 - E501 line too long (148 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 234 - E501 line too long (159 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 237 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 240 - E501 line too long (139 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 243 - E501 line too long (229 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 244 - E501 line too long (153 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 249 - E501 line too long (137 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 252 - E501 line too long (147 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 259 - E501 line too long (136 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 264 - E501 line too long (126 > 115 characters)
* corehq/apps/accounting/south_migrations/0047_update_privileges_bulk_form_management.py, line 271 - E501 line too long (136 > 115 characters)

emord commented:
FYI I merged master and moved your migration here https://github.com/dimagi/commcare-hq/commit/daeb127bdb2e53a7eebf6743f923ae221c68cc35

Commit:
ugettext_lazy

dimagimon commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -78,3 +81,146 @@ def report_context(self):
+    ARCHIVE_MODE = "archive"
+    RESTORE_MODE = "restore"
+
+    filter_options = [(ARCHIVE_MODE, ugettext_lazy('Normal Forms')), 
```
W291 trailing whitespace

emord commented:
fyi added commit @sravfeyn

sravfeyn commented:
Thanks!

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
adjust column widths

sravfeyn committed:
Add help-style option to all report-filters

sravfeyn committed:
add help-label

sravfeyn committed:
help-message after archive/restore done

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
show correct count

sravfeyn committed:
ui tweaks

sravfeyn committed:
merge remote

sravfeyn committed:
use lazy gettext for model string

sravfeyn committed:
fix ui intraction

sravfeyn committed:
refactor

sravfeyn committed:
ui

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
align buttons

sravfeyn committed:
use data_cleanup privilege

sravfeyn committed:
remove bulk-form-management privilege

sravfeyn committed:
put form-management in old-privilege

sravfeyn committed:
remove not-needed migration

sravfeyn committed:
Revert "put form-management in old-privilege"

This reverts commit 1beace80cce2873583e7e2d5fd6aca65eec1d179.

sravfeyn commented:
I had added an accounting migration (that just calls `cchq_prbac_bootstrap`) to create a role that is no longer needed. So I removed the migration and Role. Prod shouldn't have any problems, but staging might. I am planning to remove the role manually on staging and hoping migration history shouldn't have any problem either, especially, since all that these migrations do is simply call the management command. Any other gotchas here?

This leads me to the question - why do we use migrations for a simple management-command run on the accounting app?

snopoke commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -78,3 +81,156 @@ def report_context(self):
+            ] + display.other_columns
+
+    @property
+    def form_ids_response(self):
```
Where does this get called from?

snopoke replied:
I see it below, sneaky

sravfeyn replied:
I will add a note. This is not obvious

snopoke commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -78,3 +81,156 @@ def report_context(self):
+            params={'domain.exact': self.domain},
+            q=self.filters_as_es_query(),
+            es_url=XFORM_INDEX + '/xform/_search',
+        )
```
Seems like you should pass in `fields=['_id']` to avoid fetching the entire form

sravfeyn replied:
Nice! Will do

snopoke commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -450,3 +456,84 @@ def post(self, request, *args, **kwargs):
+            # Altough evaluating form_ids and sending to task would be cleaner,
+            # heavier calls should be in in an async task instead
+            import urllib
+            form_query_string = urllib.unquote(self.request.POST.get('select_all'))
```
Should you not be returning the entire query string and not just the 'select_all' parameter?

sravfeyn replied:
The value of the `select_all` param is the query_string.

snopoke commented on corehq/apps/data_interfaces/utils.py:
```diff
@@ -38,14 +40,23 @@ def add_cases_to_case_group(domain, case_group_id, uploaded_data):
+def archive_forms_old(domain, user, uploaded_data):
+    # used by excel archive forms
+    form_ids = [row.get('form_id') for row in uploaded_data]
+    return archive_or_restore_forms(domain, user, form_ids, 'archive', from_excel=True)
```
`'archive'` should be an instance of `FormManagementMode`

snopoke commented on corehq/apps/reports/display.py:
```diff
@@ -10,6 +20,50 @@ def replace(self, *args):
         return StringWithAttributes(string)
 
 
+class FormDisplay(object):
```
Nice

snopoke commented:
Looks good beyond a few comments.

sravfeyn committed:
Merge branch 'master' into FormArchive

sravfeyn committed:
PR feedback

dimagimon commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -78,3 +81,159 @@ def report_context(self):
+    ARCHIVE_MODE = "archive"
+    RESTORE_MODE = "restore"
+
+    filter_options = [(ARCHIVE_MODE, ugettext_lazy('Normal Forms')), 
```
W291 trailing whitespace

dimagimon commented on corehq/apps/data_interfaces/tasks.py:
```diff
@@ -31,3 +31,41 @@ def bulk_archive_forms(domain, user, uploaded_data):
+        raise Exception("No formids supplied")
+    response = archive_or_restore_forms(domain, couch_user, xform_ids, mode, task)
+    return response
+
```
W391 blank line at end of file

sravfeyn committed:
pep

sravfeyn commented:
Hey @snopoke, I have responded to the comments

snopoke commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -78,3 +81,159 @@ def report_context(self):
+
+    @property
+    def rows(self):
+        submissions = [res['_source'] for res in self.es_results.get('hits', {}).get('hits', [])]
```
You should make this a generator to avoid generating the list in memory

snopoke commented on corehq/apps/data_interfaces/tasks.py:
```diff
@@ -31,3 +31,40 @@ def bulk_archive_forms(domain, user, uploaded_data):
+
+    if not xform_ids:
+        # should never be the case
+        raise Exception("No formids supplied")
```
I don't think raising an exception there is necessary. If you want to know if this happens use a soft assert. Otherwise just return an empty response.

snopoke replied:
Or a response with an appropriate message

snopoke commented:
Not mergable

snopoke commented:
@sravfeyn bump

sravfeyn committed:
pr feedback

sravfeyn commented:
@snopoke (sorry didn't get time the other day)

amsagoff commented:
@biyeun for styling of the select all

sravfeyn commented:
hey @amsagoff should we hold off on merge till the UI change or can we merge and do that seperately?

Also for context, here is how it looks like now.
<img width="665" alt="screen shot 2015-09-10 at 11 43 57 am" src="https://cloud.githubusercontent.com/assets/829142/9781390/556efd9e-57b1-11e5-8d6d-cbca73a0e170.png">

amsagoff commented:
I don't think blocking, can be separate PR

czue commented:
woot!

amsagoff commented:
What is going on with this page: https://www.commcarehq.org/a/banaras/data/edit/bulk_archive_forms/

Looks like an internal name got exposed.

Did we decide to do something different with the paywall here? Seems like a cool experiment, are there analytics to go with it? @czue @snopoke @sravfeyn

sravfeyn commented:
Hey looks like this was because a title was not added. This will be fixed in the next deploy https://github.com/dimagi/commcare-hq/pull/8212/files
